modules/auxiliary/scanner/scada/modbusclient.rb

```diff
@@ -369,6 +386,31 @@ def write_registers
       print_error("Unknown answer")
     end
   end
+  
+  def read_id
+    @function_code = 0x2b
+    print_status("Sending READ ID...")
+    response = send_frame(make_read_id_payload)
+    if response.nil?
+      print_error("No answer for the READ REGISTER")
+    elsif response.unpack("C*")[7] == (0x80 | @function_code)
+      handle_error(response)
+    elsif response.unpack("C*")[7] == @function_code
+      @ObjCnt = response[13].unpack("C")[0]
```

Comment:
Same point as below here, you aren't checking the length after these `unpack` statements which could return any length of data, so its entirely possible that I send you data that meets your `unpack` query but is less than 8 bytes long, thus causing the `[7]` part to go out of bounds.